### PR TITLE
pcl: fix configurePhase with withCuda=true

### LINKED ISSUE #275090: Build failure: pcl with CUDA enabled on x86_64-linux
### Steps To Reproduce

1. Run `nix develop` with the following flake:

```nix
{
  inputs = { nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable"; };

  outputs = { self, nixpkgs }: {
    devShells = nixpkgs.lib.genAttrs [ "x86_64-linux" ] (system:
      let
        pkgs = import nixpkgs {
          inherit system;
          config = {
            allowUnfree = true;
            cudaSupport = true;
          };
        };
      in {
        default = pkgs.mkShell {
          buildInputs = [ (pkgs.pcl.override { withCuda = pkgs.config.cudaSupport; }) ];
        };
      });
  };
}
```

### Build log

https://gist.github.com/al3xtjames/64ac677568ba786ed2a1823891157d0d

### Additional context

I think this broke with the recent cudaPackages changes, but I haven't figured out the responsible commit.

### Notify maintainers

@viric
@samuela
@SomeoneSerge
@ConnorBaker   

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.61-hardened1, NixOS, 24.05 (Uakari), 24.05.20231204.2c7f3c0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos"`
 - channels(alex): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by al3xtjames**
## Description of changes

https://github.com/NixOS/nixpkgs/pull/271078 caused the configurePhase of pcl to fail when
withCuda is set to true. Fix https://github.com/NixOS/nixpkgs/issues/275090 by replacing
cudatoolkit with cudaPackages.cuda_nvcc.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 275092 --extra-nixpkgs-config '{ cudaSupport = true; allowUnfree = true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by SomeoneSerge** [comment]
Hi, and thanks! The build probably broke because https://github.com/NixOS/nixpkgs/pull/271078 made cuda-related setup hooks more strict about when they're active, and it also had a regression addressed in https://github.com/NixOS/nixpkgs/pull/273048.

Note that it's not evaluation that's broken, it's the configurePhase of the build (we'll want to update the commit message)

File: pkgs/development/libraries/pcl/default.nix [hunk 1]
```diff
@@ -47,7 +47,7 @@ stdenv.mkDerivation rec {
     libXt
   ]
   ++ lib.optionals stdenv.isDarwin [ Cocoa AGL ]
-  ++ lib.optionals withCuda [ cudatoolkit ];
+  ++ lib.optionals withCuda [ cudaPackages.cuda_nvcc ];
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/pcl/default.nix:50]: `cuda_nvcc` should go in `nativeBuildInputs` (H'm, I'm even surprised this has worked for you with it in `buildInputs`)

File: pkgs/development/libraries/pcl/default.nix [hunk 2]
```diff
@@ -18,7 +18,7 @@
 , Cocoa
 , AGL
 , OpenGL
-, withCuda ? false, cudatoolkit
+, withCuda ? false, cudaPackages
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/pcl/default.nix:21]: May I ask you to also change `false` to `config.cudaSupport`?

**Comment by SomeoneSerge** [comment]
Marking for backporting because 1) using `cudaPackages.cudatoolkit` is more or less a bug, 2) this must be broken now on `release-23.11` as well

**Comment by al3xtjames** [comment]
Thanks for the suggestions. I've updated the PR (including the commit message).

**Review by SomeoneSerge — APPROVED** [review]
```
❯ NIX_PATH= nix build --refresh -I nixpkgs=flake:github:al3xtjames/nixpkgs/pcl-cudaPackages-update -f '<nixpkgs>' --arg config '{ cudaSupport = true; cudaCapabilities = [ "8.6" ]; enableForwardCompat = false; allowUnfree = true; }' -L pcl --print-out-paths 
/nix/store/wdrpmnri6inqlh615gzsc1jf6ldhf26y-pcl-1.13.0
```

LGTM. Waiting for ofborg an merging

**Comment by SomeoneSerge** [comment]
Thanks @al3xtjames!

Unrelated to this PR, but we should probably mark the `x86_64-darwin` build as broken ([ofborg](https://logs.ofborg.org/?key=nixos/nixpkgs.275092&attempt_id=70cc3cfb-93e8-46b4-976d-e174f7413d91), [hydra](https://hydra.nixos.org/job/nixpkgs/trunk/pcl.x86_64-darwin))

**Pull request merged**